sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -441,6 +456,24 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
 
       case Not(IsNull(e)) => IsNotNull(e)
       case Not(IsNotNull(e)) => IsNull(e)
+
+      // Using Not(null) == null rules
+      case IsNull(Not(e)) => IsNull(e)
+      case IsNotNull(Not(e)) => IsNotNull(e)
```

Comment:
shall we have a more general optimization in `NullPropagation`? It seems like we can remove many expressions under `IsNull`/`IsNotNull` if they are `NullIntolerant`

Reply:
+1 to generalize them in `NullPropagation`

Reply:
Thank you. Great point. I will spend some time to think trough how I can achieve that.

Reply:
Updated